proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+    # or "login:id".
+    # Login names can only contain alphanumeric characters and
+    # non-repeating hyphens, and may not start or end with hyphens.
+    # GitHub's users are allowed to change their login name,
+    # although it is inconvenient. If a GitHub user changed
+    # their login name from "foo" to "bar", then a second user
+    # might change their name from "baz" to "foo" in order
+    # to take the old username of the first user. For this
+    # reason, it is not as safe to make authorization decisions
+    # based only on the login name.
+    # If desired, an admin could configure identity
+    # transformation expressions on a FederationDomain to
+    # further customize these usernames.
+    # Defaults to "login:id", which is the login name, followed
+    # by a colon, followed by the unique and unchanging integer
+    # ID number. This blends human-readable login names with the
+    # unchanging ID value. Note that colons are not allowed in
+    # GitHub login names or in ID numbers, so the colon in the
+    # "login:id" name will always be the login/id separator colon.
+    username: id
+    # Which property of the GitHub team record shall determine
+    # the group name in Kubernetes. Can be either "name" or "slug".
+    # Team names can contain upper and lower case characters,
+    # whitespace, and punctuation (e.g. "Kube admins!").
+    # Slug names are lower case alphanumeric characters and may
+    # contain dashes (e.g. "kube-admins").
+    # Either way, group names will always be prefixed by the
+    # GitHub org name followed by a colon (e.g. my-org:my-team).
+    # Note that colons are not allowed in GitHub org names,
+    # so the first colon in the group name will always be the
+    # org/team separator colon.
+    # If desired, an admin could configure identity
+    # transformation expressions on a FederationDomain to
+    # further customize these group names.
+    # Defaults to "slug".
+    groups: slug
+  # Optionally reject any user who attempts to authenticate
+  # unless they belong to one of these GitHub orgs.
+  # The GitHub App or GitHub OAuth App provided in the
+  # spec.client.secretName must be allowed by the org
+  # owners to view org memberships, or else the
+  # user will not be considered to belong to that org
+  # or belong to any teams within that org.
+  # When specified, users' group memberships will be
+  # filtered to include only those GitHub teams which
+  # are owned by these orgs.
+  # If desired, an admin could configure identity
+  # policy expressions on a FederationDomain to
+  # further customize which users and groups are allowed
+  # to authenticate.
+  # When not set, any GitHub user can authenticate
+  # regardless of org membership, and all the team
+  # memberships returned by the GitHub API for that
+  # user will be reflected as groups regardless of
+  # which orgs own those teams.
+  allowOrganizations: [ vmware-tanzu, broadcom ]
```

Comment:
I think we should clearly document that the user must specify `allowOrganizations`, and that the org into which the `GitHub App` is installed has no bearing on this configuration.

We could also require that `allowOrganizations` can only be empty if the user has explicitly set some other variable (perhaps `allowAllGitHubUsers: true`) as a guard to ensure that Pinniped admins must choose to allow users to login without reference to their org membership.

Reply:
Note: It's generally recommended to avoid boolean expressions, so we could use `allowUsersInAllGitHubOrgs: allow` or something similar. If `allowUsersInAllGitHubOrgs: allow` is configured, then `allowOrganizations` must have a length of zero. This could be enforced fairly easily in recent K8s versions with https://kubernetes.io/blog/2022/09/23/crd-validation-rules-beta/. Pinniped on older versions of K8s could just ignore `allowOrganizations` if `allowUsersInAllGitHubOrgs: allow` is configured.

Reply:
TODO: Add this to proposal.

Reply:
Should org membership (in one of the listed organizations here) be re-verified for every upstream group refresh?

I think yes, otherwise Pinniped could issue tokens for users who belong to groups that have been removed from this list.

Example: User X with org membership A,B logs in while `allowOrganizations: [A,B,C ]` and their downstream token lists all their teams in orgs A and B. Then an admin updates `allowOrganizations: [ C ]` after which user X needs to perform a token refresh. If the new version of this list isn't checked, I think Pinniped would issue a token with no groups (since the user does not belong to org C).

Reply:
TODO: when the names of these fields are chosen, add them to the proposal. See https://github.com/vmware-tanzu/pinniped/pull/1900#discussion_r1539523783 for naming discussion.